### PR TITLE
#516 Throw an error if there is a SVC in flat starting point mode

### PR DESCRIPTION
**Pull request by barondim**
**Checklist before requesting a review**

*use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes*

- [x] unit tests and non regression tests were added (update of inputs, outputs, SA, algos)
- [ ] main documentation was updated (update of input/output file)
- [ ] if this commit targets a release branch: the corresponding milestone was added in the ticket and in this PR

### CODE REVIEW

**Comment by barondim** [comment]
- [ ] TU are still to be added

File: tests/algo/TestAlgoDynModel.cpp [hunk 1]
```diff
@@ -38,7 +38,7 @@ TEST(TestAlgoDynModel, base) {
   using dfl::algo::DynamicModelDefinitions;
   using dfl::inputs::DynamicDataBaseManager;
 
-  DynamicDataBaseManager manager("res/setting.xml", "res/assembling.xml");
+  DynamicDataBaseManager manager("res/setting.xml", "res/assembling.xml", dfl::inputs::Configuration::StartingPointMode::WARM);
```
Comment (barondim) [inline tests/algo/TestAlgoDynModel.cpp:41]: Do we have to input warm or flat by default ?

File: sources/Inputs/src/AssemblingDataBase.cpp [hunk 2]
```diff
@@ -79,6 +79,15 @@ AssemblingDataBase::AssemblingDataBase(const boost::filesystem::path& assembling
   } catch (const xml::sax::parser::ParserException& e) {
     throw Error(DynModelFileReadError, assemblingFilePath.c_str(), e.what());
   }
+
+  if (startingPointMode == dfl::inputs::Configuration::StartingPointMode::FLAT) {
+    for (std::map<std::string, DynamicAutomaton>::const_iterator dynamicAutomatonsIt = dynamicAutomatons_.begin();
+         dynamicAutomatonsIt != dynamicAutomatons_.end(); ++dynamicAutomatonsIt) {
+      if (dynamicAutomatonsIt->second.lib == "SecondaryVoltageControlSimp") {
```
Comment (barondim) [inline sources/Inputs/src/AssemblingDataBase.cpp:86]: Is this the right way to detect the element is a SVC ?

Comment (rosiereflo) [reply sources/Inputs/src/AssemblingDataBase.cpp:86]: I would propose to add a method "SVCInAssembling" and then use it in Context or Configuration to make this test as it does not belong to the ddb

Comment (rosiereflo) [reply sources/Inputs/src/AssemblingDataBase.cpp:86]: > Is this the right way to detect the element is a SVC ?

There is a constant for the SVC library name

File: etc/Dictionaries/DFLError_en_GB.dic [hunk 3]
```diff
@@ -29,6 +29,7 @@ UnknownProperty               =     couldn't find the property %1% in assembling
 
 //------------------ Outputs -----------------------
 MissingGeneratorParameterInSettings = parameter %1% not found in settings file for generator %2%
+NoSVCInFlatStartingPointMode =        cannot use a svc in flat starting point mode
```
Comment (rosiereflo) [inline etc/Dictionaries/DFLError_en_GB.dic:32]: I would propose "simulation of secondary voltage controls is not possible with 'FLAT' starting point mode"

File: sources/Inputs/include/AssemblingDataBase.h [hunk 4]
```diff
@@ -418,7 +426,8 @@ class AssemblingDataBase {
   std::unordered_map<std::string, MultipleAssociation> multipleAssociations_;       ///< list of multiple associations
   std::unordered_map<std::string, std::string> generatorIdToSingleAssociationsId_;  ///< generator id associated to the single associations that contains it
   std::map<std::string, DynamicAutomaton> dynamicAutomatons_;                       ///< list of dynamic automatons
-  std::unordered_map<std::string, Property> properties_;                            ///< list of properties
+  bool containsSVC_;                                      ///< true if the assembling data base contains one or more SVC, false otherwise
```
Comment (rosiereflo) [inline sources/Inputs/include/AssemblingDataBase.h:429]: comment alignment

File: sources/Inputs/src/AssemblingDataBase.cpp [hunk 5]
```diff
@@ -79,6 +80,14 @@ AssemblingDataBase::AssemblingDataBase(const boost::filesystem::path& assembling
   } catch (const xml::sax::parser::ParserException& e) {
     throw Error(DynModelFileReadError, assemblingFilePath.c_str(), e.what());
   }
+
+  std::map<std::string, inputs::AssemblingDataBase::DynamicAutomaton>::const_iterator dynamicAutomatonsIt;
```
Comment (rosiereflo) [inline sources/Inputs/src/AssemblingDataBase.cpp:84]: Iterators should always be declared within the for loop

Comment (rosiereflo) [reply sources/Inputs/src/AssemblingDataBase.cpp:84]: Do we really need to loop afterward? maybe we could get this information on the fly inside the parser?

**Review by rosiereflo — CHANGES_REQUESTED** [review]

File: sources/Inputs/include/AssemblingDataBase.h (hunk 4 above)
Comment (barondim) [reply sources/Inputs/include/AssemblingDataBase.h:429]: The comments were aligned by clang-format

File: sources/Inputs/src/AssemblingDataBase.cpp (hunk 5 above)
Comment (barondim) [reply sources/Inputs/src/AssemblingDataBase.cpp]: I'm not sure it's possible because xml::sax::parser::Parser is in libxml, a dynawo 3rdParty.

Comment (rosiereflo) [reply sources/Inputs/src/AssemblingDataBase.cpp]: I meant in AssemblingXmlDocument

**Review by marcochiaramello — CHANGES_REQUESTED** [review]

File: sources/Context.cpp [hunk 6]
```diff
@@ -60,6 +60,11 @@ Context::Context(const ContextDef& def, const inputs::Configuration& config) :
     algoResults_(new algo::AlgorithmsResults()),
     jobEntry_{},
     jobsEvents_{} {
+  if (config_.getStartingPointMode() == inputs::Configuration::StartingPointMode::FLAT &&
```
[the remark this answers is not in the document]
Comment (rosiereflo) [reply sources/Context.cpp:63]: The other location where we could do it is in main:!buildContext as we need both the assembling and configuration to be available. What do you think?

File: sources/Context.h [hunk 7]
```diff
@@ -119,6 +119,14 @@ class Context {
    */
   void execute();
 
+  /**
+   * @brief returns if the assembling data base contains one or more SVCs
+   * @returns the attribute value
```
Comment (rosiereflo) [inline sources/Context.h:124]: wrong comment

File: sources/Inputs/include/AssemblingDataBase.h [hunk 8]
```diff
@@ -398,6 +398,14 @@ class AssemblingDataBase {
     return dynamicAutomatons_;
   }
 
+  /**
+   * @brief determines if the assembling data base contains one or more SVCs
+   * @returns the attribute value
```
Comment (rosiereflo) [inline sources/Inputs/include/AssemblingDataBase.h:403]: wrong comment

**Review by rosiereflo — CHANGES_REQUESTED** [review]

**Review by rosiereflo — APPROVED** [review]

**Pull request merged**